### ISSUE TITLE
Possible bug with TextInput and Chinese input method on iOS

### ISSUE THREAD

**Issue by Kulbear** [facebook/react-native#12599]
### Description

I have some TextInput components in my project and there should be some problems with `onEndEditting` and `onBlur` props of TextInput component. If I make the TextInput blur before I confirm the entered content from the Chinese input method, all the changes I made during this round of focus will be lost. Check the video below, starts from 0:17, the video demonstrates the issue.

Basically, it should be an issue between the system Chinese input method and the RN TextInput component. I made a 30-second video record to demonstrate this issue.

https://www.youtube.com/watch?v=BG0WNHW2MEc

![repro](https://cloud.githubusercontent.com/assets/14886380/23396112/f80edaac-fdcc-11e6-933a-8457a13080fa.gif)

![repro2](https://cloud.githubusercontent.com/assets/14886380/23398121/4b954c34-fdd6-11e6-9ba2-d31987a1c682.gif)

### Reproduction

I don't know how to work with Chinese input methods on rnplay, so I've pushed a sample project with minimal code for reproduction.

[Current code with a workaround](https://github.com/Kulbear/ChineseTextInputIssue)

[Original bug reproduction](https://github.com/Kulbear/ChineseTextInputIssue/tree/be0bc064f1d900f7db28d5d3a47faebdef5b537b)

I created a new project. The only modified file is the `index.ios.js`

```
import React, { Component } from 'react';
import {
  AppRegistry,
  StyleSheet,
  Text,
  TextInput,
  View
} from 'react-native';

export default class ChineseTextInputIssue extends Component {
  constructor (props) {
    super(props);
    this.state = {
      text: ''
    };
  }

  render () {
    return (
      <View style={styles.container}>
        <TextInput
          style={styles.inputField}
          placeholder="Type here to translate!"
          value={this.state.text}
          onBlur={(evt) => console.log('onBlur event:', evt.nativeEvent.text)}
          onChange={(evt) => console.log('onChange event:', evt.nativeEvent.text)}
          onChangeText={(text) => this.setState({ text })}
          onSubmitEditing={(evt) => console.log('onSubmitEditing event:', evt.nativeEvent.text)}
          onEndEditing={(evt) => console.log('onEndEditing event:', evt.nativeEvent.text)} />
        <Text style={styles.previewText}>
          {this.state.text}
        </Text>
        <TextInput style={styles.inputField} placeholder="Another to focus, meaningless" />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF'
  },
  inputField: {
    margin: 30,
    height: 50
  },
  previewText: {
    padding: 10,
    fontSize: 42
  }
});

AppRegistry.registerComponent('ChineseTextInputIssue', () => ChineseTextInputIssue);

```

### Solution

As you can see from the video, the displayed text and the text in the TextInput component should be identical.

### Additional Information

* React Native version: 0.41.2
* Platform: iOS, not tested on Android
* Operating System: macOS 10.12.3

**Comment by Kulbear** [comment]
I'll give a workaround which can fix the problem with both my original project and this reproduction project. The only modified part is inside the `render()` method. Check it below

```
render () {
  return (
    <View style={styles.container}>
      <TextInput
        style={styles.inputField}
        value={this.state.text}
        placeholder="Type here to translate!"
        onChange={(evt) => this.setState({ text: evt.nativeEvent.text })}
        onChangeText={(text) => setTimeout(() => {this.setState({ text: text })})}
        onSubmitEditing={(evt) => this.setState({ text: evt.nativeEvent.text })}
        onEndEditing={(evt) => this.setState({ text: evt.nativeEvent.text })} />
      <Text style={styles.previewText}>
        {this.state.text}
      </Text>
      <TextInput
        style={styles.inputField}
        placeholder="Another to focus, meaningless" />
    </View>
  );
}
```

Basically, `setTimeout` did the trick but I am not sure why. And I think the bug I mentioned in this issue is still worth to figure out and fix.

**Comment by Proudmoor** [comment]
thanks for your help

**Comment by shergin** [comment]
Could you please confirm that this issue still exists in current master? We recently fixed very similar issue and I hope that fix may help with this problem as well.

**Comment by Swordsman-Inaction** [comment]
@shergin Do you mean (this change)[https://github.com/facebook/react-native/blob/master/Libraries/Text/RCTTextView.m#L241]? This one doesn't fix the bug.

**Comment by Kulbear** [comment]
@shergin Will do in 1-2 days.   👏

**Comment by Kulbear** [comment]
With the dependencies below, the behavior **DOES CHANGE** but it still doesn't work as expected.
```
"dependencies": {
    "react": "~16.0.0-alpha.3",
    "react-native": "github:facebook/react-native#d5dda1b1365321c0bb64eca0d1d88909f0f6c7fd"
}
```

![repro5](https://cloud.githubusercontent.com/assets/14886380/23985528/90983de2-0a5a-11e7-8c6c-cdf8a20d98d2.gif)

**Comment by Swordsman-Inaction** [comment]
@shergin 
I dug for a while. Here is how the issue happens:
- Enter some random word
- Click outside of the `TextInput`
- The [`textFieldDidChange`](https://github.com/facebook/react-native/blob/master/Libraries/Text/RCTTextField.m#L200) gets called twice immediately, `_nativeEventCount` is increased by 2.
- The [`_onChange`](https://github.com/facebook/react-native/blob/master/Libraries/Components/TextInput/TextInput.js#L779) gets called immediately.
- [`this._inputRef.setNativeProps`](https://github.com/facebook/react-native/blob/master/Libraries/Components/TextInput/TextInput.js#L783) in `_onChange` is supposed to set native prop `_mostRecentEventCount` before the `text` is rendered, but somehow it didn't.
- So the [`eventLag`](https://github.com/facebook/react-native/blob/master/Libraries/Text/RCTTextField.m#L137) is larger than 0, and the correct `text` could not be set.(Even thought the native UITextField's text shouldn't be wrong I think).
- Shit happens....

Here is the log I got from both side:
- iOS
```
TextChange -> 196
setText -> count: 196 ; recentCount: 196, text: f
TextChange -> 197
setText -> count: 197 ; recentCount: 197, text: f g
TextChange -> 198
TextChange -> 199
setText -> count: 199 ; recentCount: 197, text: fg
setText -> count: 199 ; recentCount: 198, text: fg
```
- RN:
```
receivedCount: 196
received text:f
render
receivedCount: 197
received text:f g
render
receivedCount: 198
received text:fg
render
receivedCount: 199
received text:fg
render
render
```
The count 198 and 199 is those two change after click the outside of the TextInput.
We can see that 198 is not set to native when the correct `setText` is called, maybe it's delayed for some reason I don't know.

I cannot dig deeper since I'm not familiar with iOS(I'm an Android developer), but there are two things I don't understand:
- Why `textFieldDidChange` is called twice while using Chinese Keyboard? It won't be called while using English Keyboard.
- Why is native prop `mostRecentEventCount` not set before the text is set in `_onChange`? Normal it does.

**Comment by Swordsman-Inaction** [comment]
A simple tmp workaround is remove the condition about `eventLag`, using `if ( ![text isEqualToString:self.text])` instead of `if (eventLag == 0 && ![text isEqualToString:self.text])` in [here](https://github.com/facebook/react-native/blob/master/Libraries/Text/RCTTextField.m#L138)

But it will cause the UI syc problem when your JS running slowly, be careful if you want to use the approach.

**Comment by ethanyanjiali** [comment]
I'm having a same issue with iOS system Chinese input on TextInput.

**Comment by kidmysoul** [comment]
Have the same problem in japanese input(also in chinese input).  Only in IOS

when typing, only the first character is correctly spelled (example:　ba→ば) but the rest character does not work (example : ba→bあ)

when I use the setTimeout solution, I even cannot correctly input the first character. that solution does not workde for me.(and I' m using redux-form Field to wrap the TextInput element)

**Comment by Kulbear** [comment]
@kidmysoul My workaround seems to be invalid after several releases since I made the original post... I think this problem will happen whenever we are using an input method like Chinese or Japanese input methods that require a selection after typing some character combination...

**Comment by honeyjie** [comment]
我也遇到同样的问题，不能输入中文。。。

**Comment by Kulbear** [comment]
This is still a problem for Chinese/Japanese/Korean input method.

**Comment by wlwg** [comment]
I’m having this problem of using TextInput with Chinese Pinyin probably due to the same bug. When I set a max length in the textinput and type Pinyin in this input, the length constraint applies to the Pinyin rather than the final resulting Chinese characters.

**Comment by Kulbear** [comment]
I think this issue could potentially affect a lot of Chinese, Korean, Japanese users, or say non-alphabetic language users.

**Comment by kidmysoul** [comment]
Hey, when I update react native to version 0.54.0 ,this problem appears again ！？

liked there were lots of changes about textinput in ios in 0.54.0 and maybe that effects.

`            <TextInput
                style={styles.title}
                onChangeText={this.onTitleChange}
                maxLength={100}
                placeholder="例....."
                underlineColorAndroid={'transparent'}
                value={title}/>`

this simple code above cannot even input a whole Japanese word in ios.

**Comment by jiangbophd** [comment]
so Big Issue!!!!!

**Comment by mkjfeng01** [comment]
problem happened in react-native version 0.55.0

**Comment by gelove** [comment]
problem happened in react-native version 0.54.2

**Comment by mkjfeng01** [comment]
@gelove Problem has been solved here #18456, the code has not been merged to newest version, so you have to modify your code manually https://github.com/facebook/react-native/pull/18456/files

**Comment by AKACC** [comment]
I took out `value` from `Textinput` and then the problem fixed. It seems every letter typed would trigger `onChangeText` and render again. Wish this can help.

**Comment by ankur-sardar** [comment]
Any Update on this issue?

**Comment by lzxb** [comment]
I also have the same problem, the version is 0.55.4

**Comment by daskinnyman** [comment]
this problem still occurs in 55.3

**Comment by wujunchuan** [comment]
I have to use  the `onEndEditing` callback to replace the `onChange` callback.
But it won't be excuted timely

```jsx
<TextInput
  style={[ commonStyles.commonInput, commonStyles.pdl_normal, commonStyles.pdr_normal ]}
  underlineColorAndroid={'transparent'}
  // 中文支持问题， issues
  // https://github.com/facebook/react-native/issues/12599
  // value={this.state.remark}
  // onChangeText={( text ) => this.setState( { remark: text } )}
  // onChange={(evt) => this.setState({ remark: evt.nativeEvent.text })}
  // onChangeText={(text) => setTimeout(() => {this.setState({ remark: text })})}
  onEndEditing={(evt) => this.setState({ remark: evt.nativeEvent.text })}
  placeholder={I18n.t( Keys.eos_remark )}
  defaultValue={this.state.remark}
  autoCapitalize={'none'}/>
```

**Comment by Taphood** [comment]
This problem still occurs in RN56.0 as well.  

As AKACC mentioned if you remove "value" field from TextInput it seems to work even if you use maxlength and onChangeText for every letter typed.  

Since it is pretty common to read a value stored in Redux store as below I wish this bug to be taken care for CJK letters input.

```JVM
                                  <TextInput
                                    value={this.props.displayName}       // <--- remove this line
                                    label="some_text"
                                    onChangeText={val=> this.onChangeInput("displayName", val)}
                                    maxLength={8}
                                    placeholder="some_placeholder_text"
                                    autoCapitalize="none"
                                /> 
```

**Comment by xuechinahb** [comment]
This bug has been fixed. [https://github.com/facebook/react-native/pull/18456/files](https://github.com/facebook/react-native/pull/18456/files). I merge this code manually, it works well.

**Comment by hkung77** [comment]
Can I get some idea on when and/or which release will have the bug fix?

**Comment by kidmysoul** [comment]
@hkung77  maybe 0.57.0 
and the 0.57.0-rc has already fixed this bug. you can use it too.

**Comment by StevenMasini** [comment]
@elanpang this has been included in `0.57`, see the [changelogs](
https://github.com/react-native-community/react-native-releases/blob/master/CHANGELOG.md#ios-specific-fixes)

**Comment by elanpang** [comment]
@StevenMasini  So luck :)
